nc/source/physics/CollisionDetection.inl

```diff
@@ -0,0 +1,119 @@
+     *  Frame 2: c2 collision events incorrectly based on past c1 state
+     *  Solution: Can maybe keep handles, but more likely, we will want
+     *  ownership of colliders in physics somewhere. */
+    struct CollisionData
+    {
+        Collider* first;
+        Collider* second;
+    };
+
+    inline void DoCollisionStep(const std::vector<Collider*>& colliders);
+    inline void BuildCurrentFrameState(const std::vector<Collider*>& colliders, std::vector<CollisionData>& out);
+    inline void CompareFrameStates(const std::vector<CollisionData>& previousState, const std::vector<CollisionData>& currentState);
+    inline void NotifyCollisionEvent(const CollisionData& data, CollisionEvent type);
+    inline bool operator ==(const CollisionData& lhs, const CollisionData& rhs);
+
+    inline void DoCollisionStep(const std::vector<Collider*>& colliders)
+    {
+        static std::vector<CollisionData> previousCollisions{}; //can't remain static b/c engine can't clear it on scene change
+        std::vector<CollisionData> currentCollisions;
+        currentCollisions.reserve(previousCollisions.size());
+        BuildCurrentFrameState(colliders, currentCollisions);
+        CompareFrameStates(previousCollisions, currentCollisions);
+        previousCollisions = std::move(currentCollisions);
+    }
+
+    inline void BuildCurrentFrameState(const std::vector<Collider*>& colliders, std::vector<CollisionData>& out)
+    {
+        NC_PROFILE_BEGIN(debug::profiler::Filter::Engine);
+        DirectX::BoundingOrientedBox unit, a, b;
+        const auto count = colliders.size();
+        for(size_t i = 0u; i < count; ++i)
+        {
+            unit.Transform(a, colliders[i]->GetTransformationMatrix());
+            for(size_t j = i + 1; j < count; ++j)
+            {
+                unit.Transform(b, colliders[j]->GetTransformationMatrix());
+                if(a.Intersects(b))
+                    out.emplace_back(colliders[i], colliders[j]);
+            }
+        }
+        NC_PROFILE_END();
+    }
+    
+    inline void CompareFrameStates(const std::vector<CollisionData>& previous, const std::vector<CollisionData>& current)
+    {
+        NC_PROFILE_BEGIN(debug::profiler::Filter::Engine);
+        auto currBeg = current.cbegin();
+        auto currEnd = current.cend();
+        for(auto& prev : previous)
+        {
+            if(currEnd == std::find(currBeg, currEnd, prev))
+                NotifyCollisionEvent(prev, CollisionEvent::Exit);
+            else
+                NotifyCollisionEvent(prev, CollisionEvent::Stay);
+        }
+
+        auto prevBeg = previous.cbegin();
+        auto prevEnd = previous.cend();
+        for(auto& curr : current)
+        {
+            if(prevEnd == std::find(prevBeg, prevEnd, curr))
+                NotifyCollisionEvent(curr, CollisionEvent::Enter);
+        }
+        NC_PROFILE_END();
+    }
+
+    inline void NotifyCollisionEvent(const CollisionData& data, CollisionEvent type)
```

Comment:
This method especially and this whole file is really easy to read and understand.